### PR TITLE
Element-wise losses for multi-GPU/TPU strategy

### PR DESCRIPTION
**Pull request by cofri**
Keras losses API requires that subclasses return element-wise values, i.e. each element of a batch should return a single value. Then, a reduction operation performs either an average over the batch (default case) or a sum over the batch (useful for multi-GPU or TPU trainings).

This PR proposes a refactoring of all provided losses to fit the Keras losses API. For hinge losses, this is straightforward. For KR losses, it is more complicated: by definition, KR losses compute average over the batch. We then propose two element-wise KR formulations for standard training or multi-GPU/TPU training:
* a standard element-wise KR formulation that accepts classic one-hot encoded labels,
* a KR formulation appropriate for multi-GPU/TPU, where labels must be specifically pre-processed.

Multiple API changes are introduced:
* the KR loss is now subclassed from tf.keras Loss. Use `KR()(y_true, y_pred)` instead of `KR(y_true, y_pred)`
* the negative_KR function is removed. Use instead `HKR(alpha=0)`
* a new utils function `process_labels_for_KR_loss()` is added for label processing when using a multi-GPU/TPU strategy
* a new argument `multi_gpu` is added in all KR-based losses (KR, HKR, MulticlassKR, MulticlassHKR). When training with multi-GPU/TPU, this flag must be set to True and the labels must be processed using `process_labels_for_KR_loss()`.

### CODE REVIEW

File: tests/test_losses.py [hunk 1]
```diff
@@ -233,12 +206,246 @@ def test_multi_margin_loss(self):
...
+
+        # Compare full batch loss and mini-batches loss
+        for loss in losses:
+            loss_val_full_batch = loss(y_true, y_pred)
+
+            i1 = num_items // 2
+            i2 = i1 + num_items // 4
+            i3 = i2 + num_items // 6
+            loss_val_minibatches = 0
+            loss_val_minibatches += loss(y_true[:i1], y_pred[:i1])
```
Comment (thib-s) [inline tests/test_losses.py:350]: 👍

**Review by thib-s — CHANGES_REQUESTED** [review]

**Review by thib-s — COMMENTED** [review]
add the underscores and update the `get_config` and it will be fine; good job !

File: deel/lip/losses.py [hunk 2]
```diff
@@ -6,63 +6,99 @@
...
+            self.kr_function = kr_multi_gpu
+        else:
+            self.kr_function = partial(kr, epsilon=self.eps)
+
+    @tf.function
+    def call(self, y_true, y_pred):
+        return self.kr_function(y_true, y_pred)
+
+    def get_config(self):
+        return super(KR, self).get_config()
```
[the remark this answers is not in the document]
Comment (cofri) [reply deel/lip/losses.py:87]: Good catch !

**Review by thib-s — APPROVED** [review]

**Review by franckma31 — APPROVED** [review]
Last modification are fine

**Comment by franckma31** [comment]
well done

**Review by franckma31 — APPROVED** [review]
Well done. Can merge
